pipeline_dp/advanced_query_builder.py

```diff
@@ -0,0 +1,521 @@
+            self._contribution_bounding_level = None
+            self._budget = None
+
+        def setPrivacyUnit(self,
+                           columnNames: ColumnNames,
+                           defaultValuesToIgnore: Optional[ValuesList] = None):
+            self._privacy_unit = columnNames
+            self._default_values_to_ignore = defaultValuesToIgnore
+            return self
+
+        def setContributionBoundingLevel(
+                self, contributionBoundingLevel: ContributionBoundingLevel):
+            self._contribution_bounding_level = contributionBoundingLevel
+            return self
+
+        def setBudget(self, eps: float):
+            self._budget = EpsBudget(eps)
+            return self
+
+        def build(self):
+            return LaplaceCountSpec(
+                privacy_unit=self._privacy_unit,
+                default_values_to_ignore=self._default_values_to_ignore,
+                contribution_bounding_level=self._contribution_bounding_level,
+                budget=self._budget)
+
+
+@dataclasses.dataclass
+class GaussianCountSpec(CountSpec):
+    privacy_unit: Optional[ColumnNames]
+    default_values_to_ignore: Optional[ValuesList]
+    contribution_bounding_level: Optional[ContributionBoundingLevel]
+    budget: EpsDeltaBudget
+
+    class Builder:
+
+        def __init__(self):
+            self._privacy_unit = None
+            self._default_values_to_ignore = None
+            self._contribution_bounding_level = None
+            self._budget = None
+
+        def setPrivacyUnit(self,
+                           columnNames: ColumnNames,
+                           defaultValuesToIgnore: Optional[ValuesList] = None):
+            self._privacy_unit = columnNames
+            self._default_values_to_ignore = defaultValuesToIgnore
+            return self
+
+        def setContributionBoundingLevel(
+                self, contributionBoundingLevel: ContributionBoundingLevel):
+            self._contribution_bounding_level = contributionBoundingLevel
+            return self
+
+        def setBudget(self, eps: float, delta: float):
+            self._budget = EpsDeltaBudget(eps, delta)
+            return self
+
+        def build(self):
+            return GaussianCountSpec(
+                privacy_unit=self._privacy_unit,
+                default_values_to_ignore=self._default_values_to_ignore,
+                contribution_bounding_level=self._contribution_bounding_level,
+                budget=self._budget)
+
+
+class Query:
+    """Represents a DP query."""
+
+    def __init__(self, data, group_by_key: ColumnNames,
+                 group_by_spec: GroupBySpec, aggregations):
+        self._data = data
+        self._group_by_key = group_by_key
+        self._group_by_spec = group_by_spec
+        self._aggregations = aggregations
+
+    def run(self, test_mode: bool = False):
+        """Runs the DP query."""
+        backend = pipeline_backend.LocalBackend()
```

Comment:
Replace the implementation of `run` to `raise NotImplemented()`, the purpose of this PR is skeleton, not the implementation. "1 purpose for 1 PR" ))